here-naksha-lib-psql/src/commonMain/kotlin/naksha/psql/executors/write/DeleteFeature.kt

```diff
@@ -0,0 +1,140 @@
+        val tuple = tuple(
+            session.storage,
+            tupleNumber,
+            feature,
+            metadata(tupleNumber, feature, featureId, flags),
+            write.attachment,
+            flags
+        )
+        val newVersion = Version(previousMetadata.version.txn + 1)
+        // If hst table enabled
+        collection.history?.let { hstTable ->
+            // copy head state into hst with txn_next === txn
+            insertHeadVersionToHst(
+                hstTable = hstTable,
+                headTable = collection.head,
+                versionInHst = previousMetadata.version,
+                featureId = featureId,
+            )
+            // also copy head state into hst with txn_next === txn and action DELETED as a tombstone state
+            insertTombstoneVersionToTable(
+                destinationTable = hstTable,
+                headTable = collection.head,
+                versionInDes = newVersion,
+                flagsWithDeleted = flags,
+                featureId = featureId,
+            )
+        }
+
+        // If del table enabled, copy head state into del, with action DELETED and txn_next === txn as a tombstone state
+        collection.deleted?.let { delTable ->
+            insertTombstoneVersionToTable(
+                destinationTable = delTable,
+                headTable = collection.head,
+                versionInDes = newVersion,
+                flagsWithDeleted = flags,
+                featureId = featureId,
+            )
+        }
+
+        removeFeatureFromHead(collection, featureId)
+        return tuple
+    }
+
+
+    private fun removeFeatureFromHead(collection: PgCollection, featureId: String) {
+        collection.head.let { headTable ->
+            val quotedHeadTable = quoteIdent(headTable.name)
+            session.usePgConnection()
+                .execute(
+                    sql = "DELETE FROM $quotedHeadTable WHERE $quotedIdColumn=$1",
+                    args = arrayOf(featureId)
+                ).close()
+        }
+    }
+
+    private fun metadata(
+        tupleNumber: TupleNumber,
+        feature: NakshaFeature?,
+        featureId: String,
+        flags: Flags,
+    ): Metadata {
+        val versionTime = session.versionTime()
+        var hash = 0
+        if (feature != null) hash = hash(feature, session.options.excludePaths, session.options.excludeFn)
+        return Metadata(
+            storeNumber = tupleNumber.storeNumber,
+            version = tupleNumber.version,
+            uid = tupleNumber.uid,
+            hash = hash,
+            createdAt = versionTime,
+            updatedAt = versionTime,
+            author = session.options.author,
+            appId = session.options.appId,
+            flags = flags,
+            id = featureId,
+            type = NakshaFeature.FEATURE_TYPE
+        )
+    }
+
+    private fun insertTombstoneVersionToTable(
```

Comment:
Will adjust.
- So the `nextVersion` is redundant right? We have the version from `tupleNumber`
- The `txn_next` will have to always be set to equal txn for hst table right? I have adjusted like below, please help me check it.
`SELECT
  $1, 
  COALESCE($2, ${PgColumn.txn}),
  COALESCE($3, ${PgColumn.uid}),
  ...`
  into
  `SELECT 
  COALESCE($1, ${PgColumn.txn}),
  COALESCE($1, ${PgColumn.txn}),
  COALESCE($2, ${PgColumn.uid}),
  ...`

Reply:
It is not, you have two use cases:
- Copy HEAD state to HISTORY, only set txn_next (aka nextVersion), leave txn, uid and flags unchanged
- Copy HEAD state to HISTORY and turn into tombstone, this time txn_next=nextVersion (same as above) plus txn=tupleNumber.version, uid=tupleNumber.uid, flags=flags, so in this specific case nextVersion is somehow redundant, but necessary, to detect the two use cases, you changed query does not work for either the first or second use case, because tupleNumber is null, when txn, uid and flags need to stay unchanged